### ISSUE TITLE
Help for drawing a dynamic graph

### ISSUE THREAD

**Issue by Cestovatels** [ultralytics/ultralytics#19800]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello. 
The structure I'm trying to build is actually validation from a test data and using these metrics to extract a roc curve. However, I want to build a dynamic structure. I wrote such a code but I was not very successful. How can I make the ROC curve draw?
This is an example of single-class segmentation. There is confusion when the background comes into play.
Thanks for your answers.

```py
from ultralytics import YOLO
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, auc
from sklearn.preprocessing import label_binarize

if __name__ == "__main__":
    model = YOLO("best.pt")
    metrics = model.val(data="custom.yaml", split="test", imgsz=240)
    conf_matrix = metrics.confusion_matrix.matrix
    class_count = len(metrics.names)
   
    y_true = []
    y_score = []

    for i in range(class_count):  
        for j in range(class_count):  
            y_true.extend([i] * int(conf_matrix[i, j]))
            y_score.extend([j] * int(conf_matrix[i, j]))

    y_true = np.array(y_true)
    print(y_true)
    y_score = np.array(y_score)
    print(y_score)

    n_classes = class_count
    y_true_bin = label_binarize(y_true, classes=list(range(0, class_count + 1)))
    y_score_bin = label_binarize(y_score, classes=list(range(0, class_count + 1)))
    print(y_true_bin,y_score_bin)
    class_names = list(metrics.names.values())
    print("class_names",class_names)
    plt.figure(figsize=(10, 6))
    for i in range(n_classes):
        fpr, tpr, _ = roc_curve(y_true_bin[:, i], y_score_bin[:, i])
        roc_auc = auc(fpr, tpr)
        plt.plot(fpr, tpr, label=f'{class_names[i]} (AUC = {roc_auc:.2f})')

    plt.plot([0, 1], [0, 1], 'k--')
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title('ROC Curve')
    plt.legend(loc='lower right')
    plt.savefig("ROC.jpg")
    plt.show()
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Cestovatels, thank you for reaching out and sharing your question with us 🚀! We recommend reviewing the [Ultralytics Docs](https://docs.ultralytics.com/) for insights into validation metrics and custom usage examples. You might also find the [Python API](https://docs.ultralytics.com/usage/python/) section helpful for understanding how to extend YOLO functionality.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/). This will help us debug the issue effectively.

If this is a ❓ Custom Training/Usage Question, please share more details about your use case. For example:
- What challenges are you facing with your current approach?
- Are you observing any specific errors or unexpected behavior?

Also, ensure you’ve followed our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to optimize your model and validation processes.

## Upgrade

Please ensure you're using the latest version of the `ultralytics` package to benefit from recent improvements and bug fixes. Upgrade with:

```bash
pip install -U ultralytics
```

## Environments

To simplify your development and testing, you can run YOLO in one of our verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. Learn more in our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/).
- **Amazon** Deep Learning AMI. See the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker Image**. Follow the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

## Community and Resources

For additional support, join the Ultralytics community:
- Engage in real-time discussions on [Discord](https://discord.com/invite/ultralytics) 🎧.
- Dive into in-depth threads on [Discourse](https://community.ultralytics.com/).
- Share knowledge and learn from others on our [Subreddit](https://reddit.com/r/Ultralytics).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing and confirm the correct operation of YOLO across all platforms.

This is an automated response 😊. An Ultralytics engineer will review your query and provide additional assistance soon. Thank you for your patience and for using Ultralytics! 🚀

**Comment by glenn-jocher** [comment]
@Cestovatels for dynamic ROC curve plotting with Ultralytics YOLO, consider using our built-in metrics utilities. The `metrics.plot_pr_curve` and `metrics.ap_per_class` functions in the [ultralytics.utils.metrics](https://docs.ultralytics.com/reference/utils/metrics/) module can help visualize performance metrics. 

Your current approach using the confusion matrix counts might not capture the necessary probability scores for proper ROC analysis. Instead, collect prediction confidence scores during validation:

```python
from ultralytics import YOLO

model = YOLO("best.pt")
results = model.val(data="custom.yaml", save_json=True)  # Saves raw predictions
```

Then process the saved predictions JSON file containing confidence scores and labels for proper ROC calculation. For multi-class segmentation, ensure you're handling class probabilities correctly using one-vs-rest methodology.

For implementation details, see our [model validation documentation](https://docs.ultralytics.com/modes/val/) and [metrics reference](https://docs.ultralytics.com/reference/utils/metrics/).
